design/MultipleQueues.md

```diff
@@ -0,0 +1,83 @@
+# Multiple Queues
+
+Multiple queues is important for:
+
+- Explicit submission onto less capable queues, such as pure copy queues.
+- Explicit distribution of submissions on machines with multiple queues of the same family.
+
+## QueueFamily
+
+There are multiple families of queues.
+Availible families of queues are surfaced via `sequence<QueueFamily> Adapter.queueFamilies`.
+Support for `graphics`, `compute`, and `copy` operations are surfaced by the respective attributes on the QueueFamily.
+Queue families with `graphics` or `compute` will always have `copy`.
+There may be queue families in the future like D3D12's video decode queues, which do not support copy operations.
+If there are any families that support graphics, there is at least one family that supports both compute and graphics.
+The most capable family is always first in the `Adapter.queueFamilies` list.
+(It will either be a graphics+compute, or at least compute)
+Exposed adapters must have at least a compute family.
+Users can also create `QueueFamily`s.
+
+## Queue Creation
+
+Queues are created at Device creation time, and are exposed via `sequence<Queue> Device.queues`.
+Each element in `sequence<QueueFamily> WebGPUDeviceDescriptor.queueRequests` results in a corresponding element in `Device.queues`.
+If an user-provided QueueFamily does not match a family in `Adapter.queueFamilies`, a more capable family is used.
+If no available family can satisfy an user-provided family, device creation fails.
+
+## Synchronization
+
+Queues can have fences inserted into them, and any queue can wait on that fence to be complete.
```

Comment:
Some links:

## Metal
- [MTLEvent](https://developer.apple.com/documentation/metal/mtlevent)
- [MTLSharedEvent](https://developer.apple.com/documentation/metal/mtlsharedevent)
- [MTLCommandBuffer.encodeWaitForEvent(_:value:)](https://developer.apple.com/documentation/metal/mtlcommandbuffer/2966543-encodewaitforevent)
- [MTLCommandBuffer.encodeSignalEvent(_:value:)](https://developer.apple.com/documentation/metal/mtlcommandbuffer/2966542-encodesignalevent)

## D3D12
- [ID3D12Fence](https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/nn-d3d12-id3d12fence)
- [ID3D12CommandQueue::Signal()](https://docs.microsoft.com/en-us/windows/desktop/api/D3D12/nf-d3d12-id3d12commandqueue-signal)
- [ID3D12CommandQueue::Wait()](https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/nf-d3d12-id3d12commandqueue-wait)
- [ID3D12GraphicsCommandList::ResourceBarrier()](https://docs.microsoft.com/en-us/windows/desktop/api/d3d12/nf-d3d12-id3d12graphicscommandlist-resourcebarrier)

## Vulkan
### Fences
- [VkFence](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkFence.html)
- [vkQueueSubmit()](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkQueueSubmit.html)
- [vkGetFenceStatus()](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkGetFenceStatus.html)

### Events
- [VkEvent](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkEvent.html)
- [vkCmdSetEvent()](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkCmdSetEvent.html)
- [vkCmdWaitEvents()](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkCmdWaitEvents.html)
- [vkGetEventStatus()](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkGetEventStatus.html)
- [vkSetEvent()](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkSetEvent.html)

### Semaphores
- [VkSemaphore](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkSemaphore.html)
- [VkSubmitInfo](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkSubmitInfo.html)

Reply:
Any queue cannot wait on a fence signalled on another queue... that will crash badly in Vulkan.

For cross-queue synch you need a semaphore.